src/ecs/Query.hpp

```diff
@@ -54,57 +67,77 @@ class cevy::ecs::Query {
                     return lhs._idx == rhs._idx;
                 };
                 friend bool operator!=(iterator const &lhs, iterator const &rhs) {
-                    return lhs._idx == rhs._idx;
+                    return lhs._idx != rhs._idx;
                 };
 
             private:
-                template <size_t... Is>
-                void incr_all(std::index_sequence<Is...>) {
+                static constexpr  std::index_sequence_for<T...>_seq() {
+                    return std::index_sequence_for<T...>();
+                };
+                // template <size_t... Is>
+                void incr_all( /* std::index_sequence<Is...> */ ) {
+                    if (_idx == _max)
+                        return;
+                    do {
+                        _idx++;
+                        ([&] {
+                            (std::get<iterator_t<SparseVector<T>>>(current)++);
+                        } (), ...);
+                    } while (!all_set() && _idx < _max); // NOTE - check to choose <= or <
+                }
+
+                void sync() {
+                    if (_idx == _max)
+                        return;
                     while (!all_set() && _idx < _max) { // NOTE - check to choose <= or <
                         _idx++;
                         ([&] {
-                            (std::get<Is>(current)++);
+                            (std::get<iterator_t<SparseVector<T>>>(current)++);
                         } (), ...);
                     }
                 }
 
-                template <size_t... Is>
-                bool all_set(std::index_sequence<Is...>) {
-                    if (((std::nullopt == (*std::get<Is>(current))) && ...)) {
+                // template <size_t... Is>
+                bool all_set( /* std::index_sequence<Is...> */ ) {
+                    if (((std::nullopt == (*std::get<iterator_t<SparseVector<T>>>(current))) && ...)) {
```

Comment:
If checking for all_set the condition is invalid, do :
`return ((std::nullopt != (*std::get<iterator_t<SparseVector<T>>>(current))) && ...)`

Reply:
done